File: Objective-C/PicOnPic/ViewController.m
//
//  ViewController.m
//  PicOnPic
//
//  Created by Adam Nagy on 20/10/2012.
//  Copyright (c) 2012 Adam Nagy. All rights reserved.
//

#import "ViewController.h"
#import <MobileCoreServices/MobileCoreServices.h>

@interface ViewController ()

@end

@implementation ViewController

@synthesize popoverController;
@synthesize overlayImageView;

- (void)viewDidLoad
{
  NSLog(@"viewDidLoad");
  [super viewDidLoad];
}

- (void)didReceiveMemoryWarning
{
  NSLog(@"didReceiveMemoryWarning");
  [super didReceiveMemoryWarning];
}

- (IBAction)startCamera:(id)sender
{
  NSLog(@"startCamera");

  if (popoverController)
  {
    [popoverController dismissPopoverAnimated:YES];
    popoverController = nil;    
    return;
  }
  
  if ([UIImagePickerController isSourceTypeAvailable:
       UIImagePickerControllerSourceTypeCamera])
  {
    UIImagePickerController * imagePicker =
    [[UIImagePickerController alloc] init];
    imagePicker.delegate = self;
    imagePicker.sourceType =
    UIImagePickerControllerSourceTypeCamera;
    imagePicker.mediaTypes = [NSArray arrayWithObjects:
                              (NSString *) kUTTypeImage,
                              nil];
    imagePicker.allowsEditing = NO;
    
    
    [self presentViewController:imagePicker
                       animated:YES completion:nil];
  }
}

- (IBAction)findPic:(id)sender
{
  NSLog(@"findPic");

  if (popoverController)
  {
    [popoverController dismissPopoverAnimated:YES];
    popoverController = nil;    
    return;
  }
    
  if ([UIImagePickerController isSourceTypeAvailable:
       UIImagePickerControllerSourceTypeSavedPhotosAlbum])
  {
    UIImagePickerController * imagePicker =
    [[UIImagePickerController alloc] init];
    imagePicker.delegate = self;
    imagePicker.sourceType =
      UIImagePickerControllerSourceTypePhotoLibrary;
    imagePicker.mediaTypes = [NSArray arrayWithObjects:
                              (NSString *) kUTTypeImage,
                              nil];
    imagePicker.allowsEditing = NO;
    
    popoverController = [[UIPopoverController alloc]
      initWithContentViewController:imagePicker];
    
    popoverController.delegate = self;
    
    [popoverController presentPopoverFromBarButtonItem:sender
      permittedArrowDirections:UIPopoverArrowDirectionUp
      animated:YES];
  }
}

- (void)imagePickerController:
(UIImagePickerController *)picker
didFinishPickingMediaWithInfo:(NSDictionary *)info
{
  NSLog(@"imagePickerController");
  
  [picker dismissViewControllerAnimated:YES completion:nil];
  
  UIImage * image =
    [info objectForKey:UIImagePickerControllerOriginalImage];

  // If it's a picture selection
  if (popoverController)
  {
    [popoverController dismissPopoverAnimated:YES];
    popoverController = nil;
  
    if (overlayImageView)
      [overlayImageView removeFromSuperview];
    
    overlayImageView = [[UIImageView alloc] initWithImage:image];

    [self.imageView addSubview:overlayImageView];
  }
  // If it's taken by the camera
  else
  {
    [popoverController dismissPopoverAnimated:YES];
      popoverController = nil;
    [self.imageView setImage:image];
  }
}

- (BOOL)popoverControllerShouldDismissPopover:
(UIPopoverController *)popoverController
{
  NSLog(@"popoverControllerShouldDismissPopover");
  return YES;
}

- (void)popoverControllerDidDismissPopover:(UIPopoverController *)popoverController
{
  NSLog(@"popoverControllerDidDismissPopover");
  self.popoverController = nil;
}

- (IBAction)pinch:(UIPinchGestureRecognizer *)sender
{
  NSLog(@"pinch");

  static CGAffineTransform origTr;
  if (sender.state == UIGestureRecognizerStateBegan)
  {
    origTr = overlayImageView.transform;
  }
  else if (sender.state == UIGestureRecognizerStateChanged)
  {
    CGFloat scale = [sender scale];
    
    CGAffineTransform tr =
      CGAffineTransformConcat(
        origTr,
        CGAffineTransformMakeScale(scale, scale));
    
    [self.overlayImageView setTransform:tr];
  }
}

- (IBAction)pan:(UIPanGestureRecognizer *)sender
{
  NSLog(@"pan");

  static CGPoint prevPt;
  if (sender.state == UIGestureRecognizerStateBegan)
  {
    prevPt = [sender locationInView:self.imageView];
  }
  else if (sender.state == UIGestureRecognizerStateChanged)
  {
    CGPoint pt = [sender locationInView:self.imageView];
    
    CGAffineTransform tr =
      CGAffineTransformMakeTranslation(
        pt.x - prevPt.x,
        pt.y - prevPt.y);
    
    CGPoint cp =
      CGPointApplyAffineTransform(self.overlayImageView.center, tr);
    
    [self.overlayImageView setCenter:cp];
    
    prevPt = pt;
  }
}

- (IBAction)rotate:(UIRotationGestureRecognizer *)sender
{
  NSLog(@"rotate");

  static CGAffineTransform origTr;
  if (sender.state == UIGestureRecognizerStateBegan)
  {
    origTr = overlayImageView.transform;
  }
  else if (sender.state == UIGestureRecognizerStateChanged)
  {
    CGFloat rotation = [sender rotation];
    
    // Scale
    CGAffineTransform tr =
      CGAffineTransformConcat(
        origTr,
        CGAffineTransformMakeRotation(rotation));
    
    [self.overlayImageView setTransform:tr];
  }
}
@end

File: Swift/PicOnPic/PicOnPic/ViewController.swift
//
//  ViewController.swift
//  PicOnPic
//
//  Created by Adam Nagy on 09/11/2015.
//  Copyright © 2015 Adam Nagy. All rights reserved.
//

import UIKit
import MobileCoreServices

class ViewController : UIViewController,
  UIImagePickerControllerDelegate,
  UINavigationControllerDelegate,
  UIPopoverControllerDelegate {

  var overlayImageView: UIImageView?
  var imageFromCamera = true;
  @IBOutlet weak var imageView: UIImageView!
  
  override func viewDidLoad() {
		print("viewDidLoad")
		super.viewDidLoad()
	}
	
	override func didReceiveMemoryWarning() {
		print("didReceiveMemoryWarning")
		super.didReceiveMemoryWarning()
	}
  
  @IBAction func startCamera(sender: UIBarButtonItem) {
		print("startCamera")

		if UIImagePickerController.isSourceTypeAvailable(UIImagePickerControllerSourceType.Camera) {
			let imagePicker: UIImagePickerController = UIImagePickerController()
			imagePicker.delegate = self
			imagePicker.sourceType = UIImagePickerControllerSourceType.Camera
			imagePicker.mediaTypes = NSArray(objects: kUTTypeImage) as! [String]
			imagePicker.allowsEditing = false
			presentViewController(
        imagePicker, animated: true, completion: {
          print("startCamera - completion")
          self.imageFromCamera = true
        })
		}
  }

	@IBAction func findPic(sender: UIBarButtonItem) {
		print("findPic")
    
		if UIImagePickerController.isSourceTypeAvailable(UIImagePickerControllerSourceType.SavedPhotosAlbum) {
			let imagePicker: UIImagePickerController = UIImagePickerController()
			imagePicker.delegate = self
			imagePicker.sourceType = UIImagePickerControllerSourceType.PhotoLibrary
			imagePicker.mediaTypes = NSArray(objects: kUTTypeImage) as! [String]
			imagePicker.allowsEditing = false
      presentViewController(
        imagePicker, animated: true, completion: {
          print("findPic - completion")
          self.imageFromCamera = false
        })
		}
	}
	
	func imagePickerController(picker: UIImagePickerController, didFinishPickingMediaWithInfo info: [String : AnyObject]) {
		print("imagePickerController")
		picker.dismissViewControllerAnimated(true, completion: nil)
		let image: UIImage = (info[UIImagePickerControllerOriginalImage] as? UIImage)!
		// If it's a picture selection or
    // one from the camera
    if imageFromCamera {
      imageView.image = image
    } else {
      if (overlayImageView != nil) {
				overlayImageView!.removeFromSuperview()
			}
      
			overlayImageView = UIImageView(image: image)
			imageView.addSubview(overlayImageView!)
    }
	}
	
	@IBAction func pinch(sender: UIPinchGestureRecognizer) {
		print("pinch")
    struct funcData {
      static var origTr = CGAffineTransform()
    }
    
		if sender.state == UIGestureRecognizerState.Began {
			funcData.origTr = overlayImageView!.transform
		} else {
			if sender.state == UIGestureRecognizerState.Changed {
				let scale: CGFloat = sender.scale
				let tr: CGAffineTransform =
          CGAffineTransformConcat(funcData.origTr,CGAffineTransformMakeScale(scale,scale))
				overlayImageView!.transform = tr
			}
		}
	}
	
	@IBAction func pan(sender: UIPanGestureRecognizer) {
		print("pan")
    struct funcData {
      static var prevPt = CGPoint()
    }
    
		if sender.state == UIGestureRecognizerState.Began {
      print("pan - Began")
			funcData.prevPt = sender.locationInView(imageView)
		} else if sender.state == UIGestureRecognizerState.Changed {
      print("pan - Changed")
      let pt: CGPoint = sender.locationInView(imageView)
      let tr: CGAffineTransform = CGAffineTransformMakeTranslation(
          pt.x-funcData.prevPt.x,pt.y-funcData.prevPt.y)
      let cp: CGPoint =
        CGPointApplyAffineTransform(
          overlayImageView!.center,tr)
      overlayImageView!.center = cp
      funcData.prevPt = pt
		}
	}
	
	@IBAction func rotate(sender: UIRotationGestureRecognizer) {
		print("rotate")
    struct funcData {
      static var origTr = CGAffineTransform()
    }
    
		if sender.state == UIGestureRecognizerState.Began {
			funcData.origTr = overlayImageView!.transform
		} else {
			if sender.state == UIGestureRecognizerState.Changed {
				let rotation: CGFloat = sender.rotation
				// Scale
				let tr: CGAffineTransform =
          CGAffineTransformConcat(funcData.origTr,CGAffineTransformMakeRotation(rotation))
				overlayImageView!.transform = tr
			}
		}
	}
}

